iodata/formats/wfn.py

```diff
@@ -125,59 +125,72 @@ def _load_helper_num(lit: LineIterator) -> List[int]:
-        symbol = line[:12].strip().title()[:2]
+        # FORMAT (A8,16X,3F12.8)
+        symbol = line[:8].strip().title()[:2]
+        atcoords[atom, 0] = float(line[24:36])
+        atcoords[atom, 1] = float(line[36:48])
+        atcoords[atom, 2] = float(line[48:60])
         atnum = sym2num.get(symbol)
+        # WFN files created with AIMAll have the symbol and index combined in one word.
         if atnum is None:
             atnum = sym2num[symbol[0]]
         atnums[atom] = atnum
-        # extract atomic coordinates
-        words = line[23:].split()
-        atcoords[atom, :] = [words[0], words[1], words[2]]
     return atnums, atcoords
 
 
-def _load_helper_section(lit: LineIterator, nprim: int, start: str, skip: int,
+def _load_helper_section(lit: LineIterator, nprim: int, start: str, skip: int, step: int,
                          dtype: np.dtype) -> np.ndarray:
     """Read CENTRE ASSIGNMENTS, TYPE ASSIGNMENTS, and EXPONENTS sections."""
     section = []
     while len(section) < nprim:
         line = next(lit)
-        assert line.startswith(start)
-        words = line.split()
-        section.extend(words[skip:])
-    assert len(section) == nprim
-    return np.array([word.replace('D', 'E') for word in section]).astype(dtype)
-
-
-def _load_helper_mo(lit: LineIterator, nprim: int) -> Tuple[str, str, str, np.ndarray]:
+        if not line.startswith(start):
+            lit.error(f"Expecting line to start with '{start}'")
+        line = line[skip:]
+        while len(line) >= step:
+            section.append(dtype(line[:step].replace('D', 'E')))
+            line = line[step:]
+    if len(section) != nprim:
+        lit.error("Number of elements in section do not match 'nprim'")
+    return np.array(section, dtype=dtype)
+
+
+def _load_helper_mo(lit: LineIterator, nprim: int) -> Tuple[int, float, float, np.ndarray]:
     """Read one section of MO information."""
     line = next(lit)
-    assert line.startswith('MO')
-    words = line.split()
-    number = words[1]
-    occ, energy = words[-5], words[-1]
-    coeffs = _load_helper_section(lit, nprim, ' ', 0, float)
+    if not line.startswith('MO'):
+        lit.error("Expecting line to start with 'MO'")
+    # FORMAT (2X,5I,27X,F13.7,15X,F12.6)
+    number = int(line[2:7])
+    occ = float(line[34:47])
+    energy = float(line[62:74])
+    # FORMAT (5E16.8)
+    coeffs = _load_helper_section(lit, nprim, '', 0, 16, float)
     return number, occ, energy, coeffs
 
 
 def _load_helper_energy(lit: LineIterator) -> float:
     """Read energy."""
-    line = next(lit).lower()
-    while 'energy' not in line and line is not None:
-        line = next(lit).lower()
-    energy = float(line.split('energy =')[1].split()[0])
+    line = next(lit)
+    while 'ENERGY' not in line and line is not None:
+        line = next(lit)
+    # FORMAT (17X,F20.12)
+    # this differs between AIMPAC and GAUSSIAN so use split() to extract energy
+    energy = float(line[17:37].split()[0])
```

Comment:
Done -- see 4e902c7

Reply:
Note: Files with this quirk include iodata/test/data/he_d_orbital.wfn.

Reply:
You know, it might be that these files were generated by someone incorrectly, too.  I think they should always have the energy field terminate at column 37.  I assumed all the included .wfn files were correct, but maybe I shouldn't have. Can you review your .wfn files in iodata/test/data/ and see if they are all correct?

Reply:
@msricher I believe all included *.wfn test files are correct; it's just that the software packages that generated them use a slightly different convention when dumping *.wfn files. If I remember correctly, most of these *.wfn files were added by me or @tovrstra. However, I went down the `iodata` git history line trying to see who added them and whether they were changed at any point. What I found is that @matt-chan added all test files currently used in `iodata` tests from `horton/data/test/` to localize test files in preparation for splitting `iodata` from `horton2`; see: 2d8f854fd06f8aa4d9e94778b6f8a2d9963921b4. So, the original commits adding these `horton/data/test/*.wfn` files only exist in `horton` git history. 

Log story short, you can assume these files are "correct" and there were no mistakes in generating them; it's just that WFN writers don't strictly follow one definition!

Reply:
I've checked. All WFN files were added by Farnaz to the old horton and I remember you did it systematically. I believe most were generated with Gaussian, but I'm not sure if that is true for all WFN test files.

To simplify such discussions, it would be good to keep track of the provenance of test files, e.g. a README (maybe even per file) documenting how it was obtained.